pkgs/top-level/cuda-packages.nix

```diff
@@ -0,0 +1,142 @@
+
+  cudnnVersionForCuda = {
+    "10.0" = "7.4.2";
+    "10.1" = "7.6.3";
+    "10.2" = "7.6.5";
+  }.${cudaVersion} or "8.1.1";
+
+  ### CuTensor
+  
+  buildCuTensorPackage = callPackage ../development/libraries/science/math/cutensor/generic.nix;
+
+  cuTensorVersions = {
+    "1.2.2.5" = {
+      sha256 = "1dl9bd71frhac9cb8lvnh71zfsnqxbxbfhndvva2zf6nh0my4klm";
+    };
+    "1.3.1.3" = {
+      sha256 = "sha256-mNlVnabB2IC3HnYY0mb06RLqQzDxN9ePGVeBy3hkBC8=";
+    };
+  };
+
+  ### Cuda Toolkit Redist
+
+  # Manifest files for redist cudatoolkit
+  cudaToolkitRedistManifests = {
+    "11.4" = ../development/compilers/cudatoolkit/redist/manifests/redistrib_11.4.4.json;
+    "11.5" = ../development/compilers/cudatoolkit/redist/manifests/redistrib_11.5.2.json;
+    "11.6" = ../development/compilers/cudatoolkit/redist/manifests/redistrib_11.6.2.json;
+  };
+
+  buildCudaToolkitRedistPackage = callPackage ../development/compilers/cudatoolkit/redist/build-cuda-redist-package.nix { };
+
+  buildCudaToolkitRedistPackages = { version, manifest }: let
+    attrs = filterAttrs (key: value: key != "release_date") (importJSON manifest);
+  in mapAttrs buildCudaToolkitRedistPackage attrs;
+
+  cudaToolkitRedistPackages = if hasAttr cudaVersion cudaToolkitRedistManifests then
+    (buildCudaToolkitRedistPackages { version = cudaVersion; manifest = cudaToolkitRedistManifests.${cudaVersion}; })
+  else {};
+
+  ### Override functions
+  
+  addBuildInputs = drv: buildInputs: drv.overrideAttrs (oldAttrs: {
+    buildInputs = (oldAttrs.buildInputs or []) ++ buildInputs;
+  });
+
+in cudaToolkitRedistPackages // {
+
+  inherit addBuildInputs cudatoolkit pkgs;
+
+  ### Additional packages such as cudnn, cutensor and nccl
+
+  cudnn = buildCuDnnPackage cudnnVersions.${cudnnVersionForCuda};
+
+  cutensor = buildCuTensorPackage rec {
+    version = if cudaMajorMinorVersion == "10.1" then "1.2.2.5" else "1.3.1.3";
+    inherit (cuTensorVersions.${version}) sha256;
+    # This can go into generic.nix
+    libPath = "lib/${if cudaMajorVersion == "10" then cudaMajoMinorVersion else cudaMajorVersion}";
+  };
+
+  magma = pkgs.magma.override { inherit cudatoolkit; };
+
+  nccl = callPackage ../development/libraries/science/math/nccl { };
+
+})).overrideScope' (final: prev: {
+
+  ### Overrides to fix the components of cudatoolkit-redist
+
+  libcufile = prev.libcufile.overrideAttrs (oldAttrs: {
+    nativeBuildInputs = oldAttrs.nativeBuildInputs ++ [
+      pkgs.addOpenGLRunpath
+    ];
+    buildInputs = oldAttrs.buildInputs ++ [
+      prev.libcublas
+      pkgs.numactl
+      pkgs.rdma-core
+    ];
+    postFixup = ''
+      ln -s $out/lib/libcufile.so $out/lib/libcufile.so.0
+      addOpenGLRunpath --force-rpath {$out,$lib}/lib/lib*.so
+      for program in $out/tools/gdsio; do
+        isELF "$program" || continue
+        addOpenGLRunpath "$program"
+      done
+
+      # addOpenGLRunpath --force-rpath {$out}/tools/*
+    '';
+    autoPatchelfIgnoreMissingDeps = true;
```

Comment:
It would really, really be nice if `autoPatchelfHook` would allow only `libcuda.so` to be missing... I don't know if this is possible but I posted https://discourse.nixos.org/t/have-autopatchelfhook-ignore-only-some-missing-dependencies/18465 on discourse to track the question.

Reply:
Exactly! great

Reply:
So that's https://github.com/NixOS/nixpkgs/pull/167397. That can be fixed up once that PR is accepted.